overlord/overlord.go

```diff
@@ -209,7 +211,45 @@ func (o *Overlord) addManager(mgr StateManager) {
 	o.stateEng.AddManager(mgr)
 }
 
-func loadState(backend state.Backend, restartHandler restart.Handler) (*state.State, error) {
+func initStateFileLock() (*osutil.FileLock, error) {
+	lockFilePath := dirs.SnapStateLockFile
+	if err := os.MkdirAll(filepath.Dir(lockFilePath), 0755); err != nil {
+		return nil, err
+	}
+
+	return osutil.NewFileLockWithMode(lockFilePath, 0644)
+}
+
+func lockWithTimeout(l *osutil.FileLock, timeout time.Duration) error {
+	startTime := time.Now()
+	const retryInterval = time.Second
+	for {
+		err := l.TryLock()
+		if err != osutil.ErrAlreadyLocked {
+			return err
+		}
+
+		if time.Since(startTime) >= timeout {
+			return errors.New("timeout for state lock file expired")
+		}
+		time.Sleep(retryInterval)
+	}
+}
+
+func (o *Overlord) loadState(backend state.Backend, restartHandler restart.Handler) (*state.State, error) {
+	flock, err := initStateFileLock()
+	if err != nil {
+		return nil, fmt.Errorf("fatal: error opening lock file: %v", err)
+	}
+	o.flock = flock
+
+	logger.Noticef("Acquiring state lock file")
+	if err := lockWithTimeout(o.flock, 20*time.Second); err != nil {
```

Comment:
maybe you could extrac the timeout into a variable
```
const overlordStateLockTimeout = 20 * time.Second
...
```
I understand that in the tests, there's always a per test root dir so we will not hit the timeouts unless in a specific test that exercises this code path?

Reply:
Done. Yes, I didn't see the unit tests failing on this lock.